compiler/qsc_formatter/src/formatter.rs

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/// This module contains the Q# formatter, which can be used by calling
+/// the format function available from this module. The formatting algorithm
+/// uses the cooked and concrete tokens from the parser crate to create a
+/// token stream of the given source code string. It then uses a sliding window
+/// over this token stream to apply formatting rules when the selected tokens
+/// match certain patterns. Formatting rules will generate text edit objects
+/// when the format of the input string does not match the expected format, and
+/// these edits are returned on using the formatter.
+use qsc_data_structures::span::Span;
+use qsc_frontend::keyword::Keyword;
+use qsc_frontend::lex::{
+    concrete::{self, ConcreteToken, ConcreteTokenKind::*},
+    cooked::TokenKind::*,
```

Comment:
These `use::MyEnum::*` should be as local as possible (even if you use them in multiple functions), so that the individual variants are not mistaken with individual types or between enums. For example, from reading the code I don't know if Whitespace belongs to `TokenKind` or to `ConcreteTokenKind`. So, instead of importing them globally it is better to import them right before the match statement where they are used, like this:
```rust
use qsc_frontend::lex{concrete::ConcreteTokenKind, cooked::TokenKind};

pub fn format(code: &str) -> Vec<Edit> {
  ...
  use TokenKind::*;
  let mut edits_for_triple = match (&one, &two, &three) { ... }
}

fn apply_rules(...) -> Vec<Edit> {
  ...
  use ConcreteTokenKind::*;
  match (&left.kind, &right.kind) { ... }
}
```

Reply:
I've moved the Enum::* uses to the match statement, though I'm not sure that will help with the ambiguity you pointed out as both `ConcreteTokenKind` and `TokenKind` are used in the same match statement.